Hap Configuration Rules Doc (#1643)
kyma-project/kyma-environment-broker: docs/contributor/03-10-hyperscaler-account-pool.md

To provision clusters through Gardener using Runtime Provisioner, Kyma Environment Broker (KEB) requires a hyperscaler (GCP, Azure, AWS, etc.) account/subscription. Managing the available hyperscaler accounts is not in the scope of KEB. Instead, the available accounts are handled by Hyperscaler Account Pool (HAP).

> HAP stores credentials for the hyperscaler accounts that have been set up in advance in Kubernetes Secrets. The credentials are stored separately for each provider and tenant. The content of the credentials Secrets may vary for different use cases. The Secrets are labeled with the **hyperscaler-type**, **euAccess**, **shared** and **tenant-name** labels to manage pools of credentials for use by the provisioning process. The **hyperscaler-type** contains hyperscaler name and region information in the format `hyperscaler_type: <HYPERSCALER_NAME>[_<PLATFORM_REGION>][_<HYPERSCALER_REGION>]`, where both `_<PLATFORM_REGION>` and `_<HYPERSCALER_REGION>` are optional. The **euAccess** and **shared** labels contain boolean values and they used to divide existing pools to secrets used by EU restricted regions and secrets shared by multiple Global Accounts. The **tenant-name** label is added when the account respective for a given Secret is claimed and it is the only one not added during Secret creation. This way, the in-use credentials and unassigned credentials available for use are tracked. The content of the Secrets is opaque to HAP.

Reviewer: ```suggestion
HAP stores credentials for the hyperscaler accounts that have been set up in advance in Kubernetes Secrets. The credentials are stored separately for each provider and tenant. The content of the credentials Secrets may vary for different use cases. The Secrets are labeled with the **hyperscaler-type**, **euAccess**, **shared** and **tenant-name** labels to manage pools of credentials for use by the provisioning process. The **hyperscaler-type** contains hyperscaler name and region information in the format `hyperscaler_type: <HYPERSCALER_NAME>[_<PLATFORM_REGION>][_<HYPERSCALER_REGION>]`, where both `_<PLATFORM_REGION>` and `_<HYPERSCALER_REGION>` are optional. The **euAccess** and **shared** labels contain boolean values and they are used to divide existing pools to secrets used by EU restricted regions and secrets shared by multiple Global Accounts. The **tenant-name** label is added when the account respective for a given Secret is claimed and it is the only one not added during Secret creation. This way, the in-use credentials and unassigned credentials available for use are tracked. The content of the Secrets is opaque to HAP.
```

Reviewer: > The **euAccess** and **shared** labels contain boolean values and they are used to divide existing pools to secrets used by EU restricted regions and secrets shared by multiple Global Accounts.

This is vague. At first glance it seems that this division makes two subset of pools: these used by EU restricted regions, and those shared by multiple GA. But there are four subsets (considering two boolean values).

The Secrets are stored in a Gardener seed cluster pointed to by HAP. They are available within a given Gardener project specified in the KEB and Runtime Provisioner configuration. This configuration uses a `kubeconfig` that gives KEB and Runtime Provisioner access to a specific Gardener seed cluster, which, in turn, enables access to those Secrets.

This diagram shows the HAP workflow:

![hap-workflow](../assets/hap-flow.drawio.svg)

> Before a new cluster is provisioned, KEB queries for a Secret based on the mandatory **hyperscaler-type** and optional **tenant-name**, **euAccess** and **shared** labels.

Reviewer 2: "KEB queries for a Secret" maybe here we need to say that it depeneds which plan/region was seleected by user, then it build proper filter for secrets

Reviewer: It queries for `Secret` or `SecretBinding`?

Author: I suspect it is about SecretBindings. Because the issue requires more discussion (and a lot of changes - every old example in the doc uses "Secrets") I propose to handle it separately in https://github.com/kyma-project/kyma-environment-broker/issues/1658.

Reviewer: Have you discussed it with Technical Writers?

> Before a new cluster is provisioned, KEB queries for a Secret based on the mandatory **hyperscaler-type** and optional **tenant-name**, **euAccess** and **shared** labels. The query depeneds on plan/region rules configuration based of which KEB constructs secret bindings filter.

Reviewer: Could we explain explicitly, that KEB queries for Secret Binding which contains reference to the name of the Secret in question.

Author: Of course, could we do this in https://github.com/kyma-project/kyma-environment-broker/issues/1658#issue-2799660009?

Reviewer: The thing is that now we have in this documentation Secret used to describe actual secret resource (first sentence in line 5) and we have Secrets meaning SecretBinding resource as below:
>  The Secrets are labeled with the **hyperscaler-type**, **euAccess**, **shared** and **tenant-name** labels

Author: KEB's HAP operates on Gardener's SecretBindings only. The documentation should be changed in #1658 not to obfuscate what this PR focuses on.

    tenant-name: {TENANT_NAME}
    hyperscaler-type: "gcp_cf-sa30"
```

## Selection Rules

### Overview

> HAP evaluates a set of rules to determine what labels to use when querying secret bindings. Input to the rules consists of an SKR's plan, hyperscaler and region. By default, **hyperscalerType** label value value without any suffix (values like aws, azure etc.) is used to search for the right pool. If evaluated to true the rules modify or add labels used in the secret bindings resource query. There are four possible rules to configure:

Reviewer: ```suggestion
HAP evaluates a set of rules to determine what labels to use when querying secret bindings. Input to the rules consists of an SKR's plan, hyperscaler, hyperscaler region and platform region. By default, **hyperscalerType** label value without any suffix (e.g. aws, azure etc.) is used to search for the right pool. If evaluated to true the rules modify or add labels used in the secret bindings resource query. There are four possible rules to configure:
```

```
hap: 
  platformRegionRule: "gcp:cf-sa30"
  clusterRegionRule: ""
  sharedRule: ""
  euAccessRule: ""
```

![hap-workflow](../assets/gcp-platform-region-pool.drawio.svg)

---

As stated previously, it is possible to include multiple elements in each configuration rule. The example below shows how to configure multiple plan/region pairs for based on platformRegionRule. The same can be configured for all other rules (clusterRegionRule, sharedRule, euAccessRule). The list below allows to KEB to operate on 3 pools for gcp plan: one withouth a region, one for cf-sa30 region and one for cf_jp30 region; 2 pools for aws plan: one without a region and one for cf_eu11 region; multiple pools for openstack plan - all with a platform region specified; and one pool for azure without a platform region.

```
hap: 
  platformRegionRule: "gcp:cf-sa30;gcp:cf-jp30;aws:cf-eu11;openstack"
  clusterRegionRule: ""
  sharedRule: ""
  euAccessRule: ""
```

![hap-workflow](../assets/gcp-platform-region-pool-list.drawio.svg)

---

> Below configuration specifies azure plan as one for which to use cluster region based search. In this case, cluster provisioned in with the azure plan would require a secret binding with `hyperscalerType: azure_<CLUSTER_REGION>` label. All of the rule examples apply all of platformRegionRule, clusterRegionRule, shareRule and euAccessRule properties.

Reviewer: This is not clear:
> All of the rule examples apply all of platformRegionRule, clusterRegionRule, shareRule and euAccessRule properties.

```
hap: 
  platformRegionRule: "gcp"
  clusterRegionRule: "azure"
  sharedRule: ""
  euAccessRule: ""
```

![hap-workflow](../assets/gcp-azure-pool.drawio.svg)

---

> All the configuration rules are independent of each other from the point of view of their configuration. If the same plan/region pair appears in more than one property, then all the rules that it appears in take effect. Below configuration specifies that for all gcp clusters the search should be done based on platform region and cluster region searches. The resulting search label would be `hyperscalerType: gcp_<PLATFORM_REGION>_<CLUSTER_REGION>`.

Reviewer: Is `property` defined somewhere, or this is synonymous with rule?

Reviewer: And I guess pairs plan/region could be used only in sharedRule and euAccessRule.

```
hap: 
  platformRegionRule: "gcp"
  clusterRegionRule: "gcp"
  sharedRule: ""
  euAccessRule: ""
```

![hap-workflow](../assets/gcp-cluster-pool.drawio.svg)

---

> `sharedRule` and `euAccessRule` are idependent but in contrast to `platformRegionRule` and `clusterRegionRule` if they are both evaluated to true the search label contains two additional labels `shared: true` and `euAccess: true`. Instead of modifying the same label value `hyperscalerType` two additional labels are added to the same query. In the example below, all gcp clusters will use the same pool of secret bindings marked with labels: `hyperscalerType: gcp_<PLATFORM_REGION>_<CLUSTER_REGION>`, `shared: true`, `euAccess: true`.

Reviewer: ```suggestion
`sharedRule` and `euAccessRule` are idependent but in contrast to `platformRegionRule` and `clusterRegionRule` if they are both evaluated to true the search label contains two additional labels `shared: true` and `euAccess: true`. Two additional labels are added to the same query. In the example below, all gcp clusters will use the same pool of secret bindings marked with labels: `hyperscalerType: gcp_<PLATFORM_REGION>_<CLUSTER_REGION>`, `shared: true`, `euAccess: true`.
```

```
hap: 
  platformRegionRule: "gcp"
  clusterRegionRule: "gcp"
  sharedRule: "gcp"
  euAccessRule: "gcp"
```

![hap-workflow](../assets/all-gcp-pool.drawio.svg)

---

> The last example shows initial configuration create to mimic the current bahaviour of KEB at the time of writing the document. The configuration enforces that:

Reviewer: ```suggestion
The last example shows initial configuration created to mimic the current bahaviour of KEB at the time of writing the document. The configuration enforces that:
```

* azure, aws, gcp have their own pools of dedicated bindings.
* gcp clusters in the region cf-sa30 use the pool of secret bindings marked with labels: `hyperscalerType: gcp_cf-sa30`,

> * sap-converged-cloud clusters use the pool of secret bindings marked with labels: `hyperscalerType: openstack_<CLUSTER_REGION>` and all of the are shared.

Reviewer: Note: the plan is named `sap-converged-cloud` but the hyperscalerName is `openstack`

> * sap-converged-cloud clusters use the pool of secret bindings marked with labels: `hyperscalerType: openstack_<CLUSTER_REGION>` and all of the are shared.

Reviewer: > and all of the are shared.

these?

> * trial clusters can use one of two pool of shared secret bindings marked with labels: `hyperscalerType: azure` or `hyperscalerType: aws` (because of hardcoded mapping of trial plan to azure or aws providers) depending on the provider type.

Reviewer: ```suggestion
* trial clusters can use one of two pools of shared secret bindings marked with labels: `hyperscalerType: azure` or `hyperscalerType: aws` (because of hardcoded mapping of trial plan to azure or aws providers) depending on the provider type.
```

> * azure clusters in the region cf-ch20 and aws clusters in the region cf-eu11 have their own dedicated pool.

Reviewer: Could we add information that these use euAccess pools?

    tenant-name: {TENANT_NAME}
    hyperscaler-type: "gcp_cf-sa30"
```

## Selection Rules

### Overview

> HAP evaluates a set of rules to determine what labels to use when querying secret bindings. Input to the rules consists of an SKR's plan, hyperscaler and region. By default, only unchanged hyperscaler type value (values like aws, azure etc.) is used to search for the right pool.  If evaluated to true the rules modify or add labels used in the secret bindings resource query. There are four possible rules to configure:

Author: I wrote hyperscaler type because of https://github.com/kyma-project/kyma-environment-broker/blob/main/common/hyperscaler/hyperscaler_type.go structure that produces the correct name.

* `hap.platformRegionRule` - if evaluated to true the `_<PLATFOR_REGION>` is appended to the `hyperscaler-type` label when searching, refered to as platform region based search. 

> * `hap.clusterRegionRule` - if evaluated to true the `_<HYPERSCALER_REGION>` is appended to the `hyperscaler-type` label when searching, refered to as cluster region based search.

Reviewer: Would it be convenient to specify what if two types of region are in play?
From BNF notation I know how it is supposed to look like, but nevertheless example could be easier to understand.

Author: Let's discuss this modification together.

* `hap.sharedRule` - if evaluated to true the `shared: true` label is used when searching, refered to as shared based search.
* `hap.euAccessRule` - if evaluated to true the `shared: true` label is used when searching, refered to as euAccess based search.
The configuration is done by specifying above rules in KEB helm values.

> Rules are independent of each other. Each plan and region combination can occur in the all rules specified above at the same time.

Author: The rules mentioned above make it possible to configure a rule with a plan only so it is not necessary to introduce additional constraints for sharedRules only

> Rules are independent of each other. Each plan and region combination can occur in the all rules specified above at the same time.

Reviewer: If rules are independent these are not mutually exclusive.

> Rules are independent of each other. Each plan and region combination can occur in the all rules specified above at the same time.

Reviewer: I am lost. I think we need proper discussion again.

> Rules are independent of each other. Each plan and region combination can occur in the all rules specified above at the same time.

Reviewer: @PK85 How your comment is related to https://github.com/kyma-project/kyma-environment-broker/issues/1520#issuecomment-2516633561

> Additional ACs:

> it could not be only per plan, we need to support that even per region. Like cf-sa30 or eu regions, they use shared pools too.

> Rules are independent of each other. Each plan and region combination can occur in the all rules specified above at the same time.

Reviewer 2: @jaroslaw-pieszka this comment is invalid. I thought that sa30 and eu use shared pools. They use dedicated ones. Deleting above comment.

### Format

Each rule consists of a semicolon separated list of plans that the rule applies to. Additionally, a plan can be extended with `:region` suffix that makes the rule evaluate to true only if a cluster is provisioned in the specified region. List entries comply with the format `<PLAN_ID_1>:<REGION_ID_1>;<PLAN_ID_2>:<REGION_ID_2>`. Either plan or region (but never both) can be specified as wildcard `*` meaning all plans or regions should apply for specific second value. The following example lists valid and invalid configuration values:
* `trial` - valid, rule is evaluated to true for the `trial` plan in all regions,

> * `trial:*` - valid, rule is evaluated to true for the `trial` plan in all regions,

Reviewer 2: I think we can remove:
- "plan:*" cause we can just put "plan"
- "*:*" cause we can use * (means all plans)

* `trial:eu` - valid, rule is evaluated to true for the `trial` plan in the `eu` region,

* `*:eu` - valid, rule is evaluated to true for all plans in the `eu` region,
* `eu:*` - invalid, plan must be specified in the first part of `<PLAN_ID>:<REGION_ID>` pair,
* `*:*` - invalid, at least one of the values must be specified in the `<PLAN_ID>:<REGION_ID>` pair.
* `*:eu;trial:eu` - valid, rule is evaluated to true for all plans in the `eu` region and for the `trial` plan in the `eu` region, configuration can be duplicated
* `trial:eu;trial:us` - valid, rule is evaluated to true for trials plans but only in `eu` and `us` regions,

### Validation

Rules validation is done during the KEB startup. If the configuration is invalid, KEB will not start and an error message will be displayed in the logs. The constraints used for validation include:
* Rules format check - all the rules must comply with the format specified above.

> * Plan existence check.

Reviewer: Is that all?
Region existence check?
Self-contradictory check?

Author: @PK85 what configuration checks do you want to have in KEB?

Reviewer 2: @ralikio Im not sure if such accurate validation is needed, but if yes then, we must check if:
- rules format (but here not every rule should have same syntax check, I commented about that in another comment)
- plan exists
- platformRegion exists
- clusterRegion exists

### Examples

The section provides configuration examples of secret bindings selection rules and shows what hyperscaler pools they correspond to.

> By default, KEB uses the `hyperscaler-type` label to search for secrets without applying any sufixes to the label's value. It means that without any configuration the search is done by simple hyperscaler type value implemented in [KEB](https://github.com/kyma-project/kyma-environment-broker/blob/main/common/hyperscaler/hyperscaler_type.go).

Reviewer: It is nit-picking but KEB does not search for secrets. It searches for secrets' names.

Author: https://github.com/kyma-project/kyma-environment-broker/pull/1643/commits/421df91adf829e8b0ae189bf9ea004b1adb1dacc

```
hap: 
  platformRegionRule: ""
  clusterRegionRule: ""
  sharedRule: ""
  euAccessRule: ""
```

![hap-workflow](../assets/default-pool.drawio.svg)

---

> The following configuration is an example of a rule that appends the platform region to the `hyperscaler-type` label when searching for secrets for all gcp clusters. The configuration means that there would be no possibillity to configure a pool with label `hyperscalerType: gcp` since KEB would search for `hyperscalerType: gcp__<PLATFOR_REGION>` bindings only.

Reviewer: ```suggestion
The following configuration is an example of a rule that appends the platform region to the `hyperscaler-type` label when searching for secrets for all gcp clusters. The configuration means that there would be no possibillity to configure a pool with label `hyperscalerType: gcp` since KEB would search for `hyperscalerType: gcp_<PLATFORM_REGION>` bindings only.
```

Reviewer: We know that secrets and bindings here refer to the same entity, but would the reader know?

```
hap: 
  platformRegionRule: "gcp"
  clusterRegionRule: ""
  sharedRule: ""
  euAccessRule: ""
```

> ![hap-workflow](../assets/gcp-platform-pool.drawio.svg)

Reviewer: What is the purpose of crossed out pools/rectangles? IMHO it is not needed.
If you want to specify everything that could not be reached/searched for/configured it won't be comprehensible.

Author: To point out what is the difference from the last example and what generic (pools with provider only) are available.

---

> It is possible to extend the configuration rule entry with region specified. Below configuration means that if a cluster is provisioned in cf-sa30 subaccount regiont then KEB would search for secret bindings with `hyperscalerType: gcp_cf-sa30` label. In other cases, for example when probisioning a gcp cluster in other region, KEB would search for `hyperscalerType: gcp` label. Differently than in previous example, in this case, it is possible to configure two different pools for the same hyperscaler type.

Reviewer: ```suggestion
It is possible to extend the configuration rule entry with region specified. Below configuration means that if a cluster is provisioned in cf-sa30 subaccount region then KEB would search for secret bindings with `hyperscalerType: gcp_cf-sa30` label. In other cases, for example when probisioning a gcp cluster in other region, KEB would search for `hyperscalerType: gcp` label. Differently than in previous example, in this case, it is possible to configure two different pools for the same hyperscaler type.
```

```
hap: 

>   platformRegionRule: "gcp:cf-sa30"

Reviewer: How to specify this rule for two (or more) platform regions?

  clusterRegionRule: ""
  sharedRule: ""
  euAccessRule: ""
```

![hap-workflow](../assets/gcp-platform-region-pool.drawio.svg)

---

As stated previously, it is possible to include multiple elements in each configuration rule. The example below shows how to configure multiple plan/region pairs for based on platformRegionRule. The same can be configured for all other rules (clusterRegionRule, sharedRule, euAccessRule). The list below allows to KEB to operate on 3 pools for gcp plan: one withouth a region, one for cf-sa30 region and one for jp30 region; 2 pools for aws plan: one without a region and one for eu11 region; multiple pools for openstack plan - all with a platform region specified; and one pool for azure without a platform region.

```
hap: 

>   platformRegionRule: "gcp:cf-sa30;gcp:cf-jp30;aws:cf-eu11;openstack"

Reviewer: In the rules do we use plan name?

>   platformRegionRule: "gcp:cf-sa30;gcp:cf-jp30;aws:cf-eu11;openstack"

Author: Correct.

```
hap: 
  platformRegionRule: "gcp"
  clusterRegionRule: "gcp"
  sharedRule: "gcp"
  euAccessRule: "gcp"
```

![hap-workflow](../assets/all-gcp-pool.drawio.svg)

---

The last example shows initial configuration create to mimic the current bahaviour of KEB at the time of writing the document. The configuration enforces that:
* azure, aws, gcp have their own pools of dedicated bindings.
* gcp clusters in the region cf-sa30 use the pool of secret bindings marked with labels: `hyperscalerType: gcp_cf-sa30`,
* sap-converged-cloud clusters use the pool of secret bindings marked with labels: `hyperscalerType: openstack_<CLUSTER_REGION>` and all of the are shared.

> * trial clusters can use one of two pool of shared secret bindings marked with labels: `hyperscalerType: azure` or `hyperscalerType: aws` depending on the provider type.

Reviewer 2: It is worthwhile to mention that that trial and provider type mappings are hardcoded in the code. To make it clear, the trial plan is treated differently than other plans.

Author: It is hard for me to find a place in that document for that remark. Modification of the commented line might obscure its relation the the configuration example. Other places in the document are not suitable as well because this is not a document about plans but about HAP. However, I have added a note about this in https://github.com/kyma-project/kyma-environment-broker/pull/1643/commits/6bab7851882a8e0cff3f8170e1168362d8ec860c

* azure clusters in the region cf-ch20 and aws clusters in the region cf-eu11 have their own dedicated pool.
```
hap: 
  platformRegionRule: "gcp:cf-sa30"
  clusterRegionRule: "sap-converged-cloud"
  sharedRule: "trial;sap-converged-cloud"
  euAccessRule: "azure:cf-ch20;aws:cf-eu11"
```

> ![hap-workflow](../assets/initial-pools.drawio.svg)

Reviewer 2: In a diagram for azure and aws shared, please mark the rectangle with small text "trial" in the corner. Cause I needed to think from where they come, ot just make it easy to understand, that they exists for trial plan.

Author: https://github.com/kyma-project/kyma-environment-broker/pull/1643/commits/76ba76a63745a35f189913d1b22a6eb954018fdf

Reviewer: But now we have pool (hyperscaler: azure, shared: true) with `trial` and the same pool without `trial`. Since we differentiate by label it is not possible.

    tenant-name: {TENANT_NAME}
    hyperscaler-type: "gcp_cf-sa30"
```

## Secret Bindings Selection Rules

<!-- rules Overview -->
HAP evaluates a set of rules to determine what labels to use when querying secret bindings. Input to the rules consists of an SKR's plan, hyperscaler and region. If evaluated to true the rules modify or add labels used in the secret bindings resource query. There are four possible rules to configure:

> * `hap.platformRegionRule` - if evaluated to true the `_<PLATFOR_REGION>` is appended to the `hyperscaler-type` label when searching, refered to as platform region based search,

Reviewer: Are `region` rules mutually exclusive?

Author: We agreed that all the rules can be inclusive because they are independent of each other.

Reviewer: So we need to put it here explicitly and possibly add an example,

Author: Done. Please see the newest changes.

> * `hap.clusterRegionRule` - if evaluated to true the `_<HYPERSCALER_REGION>` is appended to the `hyperscaler-type` label when searching, refered to as cluster region based search,

Reviewer: If we use (as we discussed `HYPERSCALER_REGION` the rule should be renamed as well to`hap.hyperscalerRegionRule`.

Author: Lets put it on hold until we decide if we want to use  cluster or hyperscaler regions.

* `hap.sharedRule` - if evaluated to true the `shared: true` label is used when searching, refered to as shared based search,

> * `hap.euAccessRule` - if evaluated to true the `shared: true` label is used when searching, refered to as euAccess based search.

Reviewer: Are sharedRule and euAccessRule mutually exclusive?

Author: We agreed that all the rules can be inclusive because they are independent of each other.

Reviewer: So we need to put it here explicitly and possibly add an example.
This means that `shared` and `euAccess` can coexist.

Reviewer: ```
platformRegionRule='trial:cf-eu11'
clusterRegionRule='trial:eu-central-1'
sharedRule='trial:eu-central-1,eu-central-2'
euAccessRule='cf-eu11'
```

Author: No, the rules is not valid because `sharedRule='trial:eu-central-1,eu-central-2' contains two regions for one plan, coma is not used for separating entries and eu-central-2 is left alone.

Reviewer: So how we can specify list of respective regions? `trial:eu-central-1, trial:eu-central-2`?

Author: Discussed, ok.

The configuration is done by specifying above rules in KEB helm values. 

> 

Reviewer 2: provide couple of examples, including this one:
```
hap: 
  sharedRule: "trial;sap-converged-cloud"
  euAccessRule: "azure:cf-ch20;aws:cf-eu11"
  clusterRegionRule: "sap-converged-cloud"
  platformRegionRule: "gcp:cf-sa30"
```

<!-- rules format -->

> Each rule consists of a semicolon separated list of plans that the rule applies to. Additionally, a plan can be extended with `:region` suffix that makes the rule evaluate to true only if a cluster is provisioned in the specified region. List entries comply with the format `<PLAN_ID_1>:<REGION_ID_1>;<PLAN_ID_2>:<REGION_ID_2>`. Either plan or region (but never both) can be specified as wildcard `*` meaning all plans or regions should apply for specific second value. The following example lists valid and invalid configuration values:

Reviewer: So `:region` suffix refers to `hyperscalerRegionRule`?

Author: It depends on the rule.  The section that you are describing is about general rule format. The same format applies to all configurations.

Reviewer 3: why the expression `*:*` is prohibited? let's say we want to use shared subscriptions for all plans/regions? Just a question, @PK85 what do you think?

Reviewer 2: @piotrmiskiewicz we need to think what it means for given rule, rules combinations, example:
```
hap: 
  platformRegionRule: "*:*"
  clusterRegionRule: "*:*"
  sharedRule: "*" // here regions doesn't make sense, I left another comment about that
  euAccessRule: "*:*"
```

It means every SKR will be shared=true, euAccess=true and hyperscaler-type: [plan]_[platformRegion]_[clusterRegion]

another example:
hap: 
  platformRegionRule: "*:*"
  clusterRegionRule: "*:*"
  sharedRule: "gcp" // here regions doesn't make sense, I left another comment about that
  euAccessRule: "*:*"
```
That means only gcp is shared and has shared=true, euAccess=true and hyperscaler-type: gcp_[platformRegion]_[clusterRegion]

and all other SKR has dedicated secretBinding and euAccess=true and hyperscaler-type: [plan]_[platformRegion]_[clusterRegion]

It's a bit hardcore but we can allow that :)

Reviewer 2: I think we should not use :region suffix in the "sharedRule" and "euAccessRule" (for euAccessRule we could allow but clearly document that this is platformRegion only allowed). Only allowed to be used in the "platformRegionRule" and or "clusterRegionRule"

> * `trial` - valid, rule will be evaluated to true for the `trial` plan in all regions,

Reviewer: This concerns only `hyperscalerRegionRule`, doesn't it?

Author: This concerns all the rules.

Reviewer: So this is not true:
> makes the rule evaluate to true only if a cluster is provisioned in the specified region

Author: It depends on the context of the given property. In the section you refer to the region is used as either platform or cluster region. Which region is used for given property is described in the `Overview` section.

Reviewer: If this concerns platform region this
> makes the rule evaluate to true only if a cluster is provisioned in the specified region
is not true.

Author: Apparently, there is a misunderstanding. Let's have a call to discuss where the ambiguity comes from.

* `trial:*` - valid, rule will be evaluated to true for the `trial` plan in all regions,

> * `trial:eu` - valid, rule will be evaluated to true for the `trial` plan in the `eu` region,

Reviewer: Will it refer to global KEB function (predicate) determining if euAccess is in place?

Author: I wanted to implement it as predicate using the configuration. But what is the point of your question? Do you want to add anything to the documentation about this?

> * `*:eu` - valid, rule will be evaluated to true for all plans in the `eu` region,

Reviewer 2: I think this is not needed, I have no usecase for that

Author: Removed.

* `eu:*` - invalid, plan must be specified in the first part of `<PLAN_ID>:<REGION_ID>` pair,
* `*:*` - invalid, at least one of the values must be specified in the `<PLAN_ID>:<REGION_ID>` pair.

> * `*:eu;trial:eu` - valid, rule will be evaluated to true for all plans in the `eu` region and for the `trial` plan in the `eu` region, configuration can be duplicated

Reviewer: What does it mean that configuration can be duplicated?

> `*:eu;trial:eu`
> `*:eu;trial:eu`

Or you mean that rule parts can define overlapping sets?

Author: Overlapping, corrected in https://github.com/kyma-project/kyma-environment-broker/pull/1643/commits/5637db35dbbe17df031cb6820b7e48442e62988e

> * `trial:eu;trial:gcp` - valid, rule will be evaluated to true for trials plans but only in `eu` and `gcp` regions,

Reviewer: So it seems that in the previous line `;` is treated as `or` operator, and in this line as `and` operator.

Reviewer: Does it mean that we will have predefined sets of regions that user can use in configuration?
`eu`, `gcp`?

Author: Corrected `gcp` to `us`. In this line "and" is not used as an operator but as a conjunction.

Reviewer: Still what `us` means? Is it predefined set of regions?

Author: It is a made up region relative only in the context of the section.

<!-- search construction -->
![alt text](image.png)

<!-- TODO: rules validation -->
Rules validation is done during the KEB startup. If the configuration is invalid, KEB will not start and an error message will be displayed in the logs. The constraints used for validation include:
* Rules format check - all the rules must comply with the format specified above.

> * Plan ?and region? existence check.

Reviewer: It seems it needs rephrasing.

Author: I was wondering if we have simple means to implement region check. I remove `?and region?` from the doc for now.
